File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    # def subsets(self, nums: List[int]) -> List[List[int]]:

    # class Solution:
    def subsets(self, nums : List[int]) -> List[List[int]]:
        res = []
        
        def backtrack(start, path):
            res.append(path[:])  # store current subset
            
            for i in range(start, len(nums)):
                path.append(nums[i])          # choose
                backtrack(i + 1, path)       # explore
                path.pop()                   # un-choose
        
        backtrack(0, [])
        return res
        

File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let result = []
        let backtrack = (path, target, start) => {
            if (0 == target){
                result.push([...path])
                return 
            }

            //  if (target < 0) return

            for (let i = start; i < nums.length && target > 0; i++){
                path.push(nums[i])
                backtrack(path, target - nums[i], i)
                path.pop()
            }

        }
        backtrack([], target, 0)

        console.log(result)

        return result
    }
}
